prov/gni/src/gnix_tags.c

```diff
@@ -0,0 +1,564 @@
+ * @return NULL, if no match is found,
+ *         a non-NULL value, if a match is found
+ */
+static inline struct gnix_tag_list_element *__tag_list_peek_first_match(
+		struct gnix_tag_storage *ts,
+		uint64_t tag,
+		uint64_t ignore,
+		struct slist *list,
+		uint64_t flags,
+		void *context,
+		struct gnix_address *addr)
+{
+	struct slist_entry *current;
+	struct gnix_tag_search_element s_elem = {
+			.tag = tag,
+			.ignore = ignore,
+			.flags = flags,
+			.context = context,
+			.use_src_addr_matching = ts->attr.use_src_addr_matching,
+			.addr = addr,
+	};
+
+	/* search the list for a matching element. stop at the first match */
+	for (current = list->head; current != NULL; current = current->next) {
+		if (ts->match_func(current, &s_elem))
+			return (struct gnix_tag_list_element *) current;
+	}
+
+	return NULL;
+}
+
+/**
+ * @brief finds and removes the first match in a tag list
+ *
+ * @param ts           pointer to gnix_tag_storage_object
+ * @param tag          tag to find
+ * @param ignore       bits to ignore in tags
+ * @param list         slist to search
+ * @param flags        fi_tagged flags
+ * @param context      fi_context associated with tag
+ * @param addr         gnix_address to find
+ * @param addr_ignore  bits to ignore in address
+ * @return NULL, if no match is found,
+ *         a non-NULL value, if a match is found
+ */
+static inline struct gnix_tag_list_element *__tag_list_find_element(
+		struct gnix_tag_storage *ts,
+		uint64_t tag,
+		uint64_t ignore,
+		struct slist *list,
+		uint64_t flags,
+		void *context,
+		struct gnix_address *addr)
+{
+	struct gnix_tag_search_element s_elem = {
+			.tag = tag,
+			.ignore = ignore,
+			.flags = flags,
+			.context = context,
+			.use_src_addr_matching = ts->attr.use_src_addr_matching,
+			.addr = addr,
+	};
+
+	/* search the list for a matching element. stop at the first match */
+	return (struct gnix_tag_list_element *)
+			slist_remove_first_match(list,
+					ts->match_func, &s_elem);
+}
+
+/**
+ * @brief checks attributes for invalid values
+ *
+ * @param attr  attributes to be checked
+ * @return -FI_EINVAL, if attributes contain invalid values
+ *         FI_SUCCESS, otherwise
+ */
+static inline int __check_for_invalid_attributes(
+		struct gnix_tag_storage_attr *attr)
+{
+	if (attr->type < 0 || attr->type >= GNIX_TAG_MAXTYPES)
+		return -FI_EINVAL;
+
+	return FI_SUCCESS;
+}
+
+int _gnix_tag_storage_init(
+		struct gnix_tag_storage *ts,
+		struct gnix_tag_storage_attr *attr,
+		int (*match_func)(struct slist_entry *, const void *))
+{
+	int ret;
+	struct gnix_tag_storage_attr *attributes = &default_attr;
+
+	if (ts->state == GNIX_TS_STATE_INITIALIZED) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+				"attempted to initialize already active tag storage\n");
+		return -FI_EINVAL;
+	}
+
+	if (attr)
+		attributes = attr;
+
+	if (__check_for_invalid_attributes(attributes)) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+				"invalid attributes passed in to init\n");
+		return -FI_EINVAL;
+	}
+
```

Comment:
Small nit.  There's no need to check if `attr` is not passed in, so I'd move this check inside the previous conditional.